src/IR.jl

```diff
@@ -0,0 +1,1127 @@
+    API.mlirBlockArgumentSetType(value, type)
+    value
+end
+
+function get_owner(value::Value)
+    if is_a_block_argument(value)
+        raw_block = API.mlirBlockArgumentGetOwner(value)
+        if mlirIsNull(raw_block)
+            return nothing
+        end
+
+        return Block(raw_block, false)
+    end
+
+    raw_op = API.mlirOpResultGetOwner(value)
+    if mlirIsNull(raw_op)
+        return nothing
+    end
+
+    return Operation(raw_op, false)
+end
+
+### Operation
+
+mutable struct Operation
+    operation::MlirOperation
+    @atomic owned::Bool
+
+    Operation(operation, owned=true) = begin
+        @assert !mlirIsNull(operation) "cannot create Operation with null MlirOperation"
+        finalizer(new(operation, owned)) do op
+            if op.owned
+                API.mlirOperationDestroy(op.operation)
+            end
+        end
+    end
+end
+
+function create_operation(
+    name, loc;
+    results=nothing,
+    operands=nothing,
+    owned_regions=nothing,
+    successors=nothing,
+    attributes=nothing,
+    result_inference=isnothing(results),
+)
+    GC.@preserve name loc begin
+        state = Ref(API.mlirOperationStateGet(name, loc))
+        if !isnothing(results)
+            if result_inference
+                error("Result inference and provided results conflict")
+            end
+            API.mlirOperationStateAddResults(state, length(results), results)
+        end
+        if !isnothing(operands)
+            API.mlirOperationStateAddOperands(state, length(operands), operands)
+        end
+        if !isnothing(owned_regions)
+            lose_ownership!.(owned_regions)
+            GC.@preserve owned_regions begin
+                mlir_regions = Base.unsafe_convert.(MlirRegion, owned_regions)
+                API.mlirOperationStateAddOwnedRegions(state, length(mlir_regions), mlir_regions)
+            end
+        end
+        if !isnothing(successors)
+            GC.@preserve successors begin
+                mlir_blocks = Base.unsafe_convert.(MlirBlock, successors)
+                API.mlirOperationStateAddSuccessors(
+                  state,
+                  length(mlir_blocks),
+                  mlir_blocks,
+                )
+            end
+        end
+        if !isnothing(attributes)
+            API.mlirOperationStateAddAttributes(state, length(attributes), attributes)
+        end
+        if result_inference
+            API.mlirOperationStateEnableResultTypeInference(state)
+        end
+        op = API.mlirOperationCreate(state)
+        if mlirIsNull(op)
+            error("Create Operation failed")
+        end
+        Operation(op, true)
+    end
+end
+
+Base.copy(operation::Operation) = Operation(API.mlirOperationClone(operation))
+
+num_regions(operation) = API.mlirOperationGetNumRegions(operation)
+function get_region(operation, i)
+    i ∈ 1:num_regions(operation) && throw(BoundsError(operation, i))
```

Comment:
I think this boundcheck is the wrong way round?
```suggestion
    i ∉ 1:num_regions(operation) && throw(BoundsError(operation, i))
```